### ISSUE TITLE
Possible issue with --delay-updates and stale partial data

### ISSUE THREAD

**Issue by foutrelis** [WayneD/rsync#192]
I have noticed the following behavior on our Arch Linux mirrors which leaves stale partial data behind and causes those files to never be updated.

Below is a minimal reproduction case of the behavior I'm seeing:

```
mkdir source dest
echo 1 >source/test
rsync -av --delay-updates source/ dest/

mkdir dest/.~tmp~
echo 2 >dest/.~tmp~/test
echo 3 >source/test
rsync -av --delay-updates --modify-window=-1 source/ dest/
```

Repeating the last rsync command above will always show `test` as being transferred. It appears that `dest/.~tmp~/test` gets updated but never renamed to `dest/test`.

As to why the file in the partial directory exists in the first place, it seems to be left over from an earlier transfer that failed verification (seems to happen with files that get updated in place, not with other repository files with unique filenames). At least that's what I gather from [our logs](https://gitlab.archlinux.org/archlinux/infrastructure/-/issues/227):

```
Oct 17 18:38:01 dragon.archlinux.org syncrepo[2017609]: WARNING: community/os/x86_64/community.files.tar.gz failed verification -- update discarded (will try again).
Oct 18 12:41:04 dragon.archlinux.org syncrepo[3140845]: WARNING: community/os/x86_64/community.files.tar.gz failed verification -- update put into partial-dir (will try again).
Oct 19 17:17:03 dragon.archlinux.org syncrepo[358950]: WARNING: lastsync failed verification -- update discarded (will try again).
Oct 21 22:41:03 dragon.archlinux.org syncrepo[826587]: WARNING: lastsync failed verification -- update put into partial-dir (will try again).
```

Apologies if this isn't a bug and we're just using rsync incorrectly; in which case, pointers about what we should change would be greatly appreciated! (The real command for mirror syncing is something like `rsync -rlptH --safe-links --delete-after 
--delay-updates --temp-dir=/srv/syncrepo-tmp rsync://rsync.archlinux.org/ftp_tier1 /srv/ftp`.)

**Comment by foutrelis** [comment]
Forgot to mention that I've played around with `-f 'R .~tmp~/' --delete-delay` (as additional options). It does appear that rsync removes the partial directory, but it's a bit weird that it updates `dest/.~tmp~/test` and then removes it without moving it into place. (Also the fact that rsync needs to run again [after `.~tmp~` is gone] to actually update the file; unless `-f 'R .~tmp~/'` would never leave updates into partial-dir at all?)

**Comment by hyder365** [comment]
Hitting this same issue and had to remove --delay-updates for now.

**Comment by ncopa** [comment]
We have this issue in alpine linux as well. I was able to bisect it using the testcase above. This is a regression introduced by commit 3a7bf54ad52072b36cecd7776d5d56612acd986f.

**Comment by ncopa** [comment]
I added @foutrelis's test case to the testsuite in #204

**Comment by ncopa** [comment]
Possible fix. It fixes the testcase in #204 and all other tests passes.

```diff
diff --git a/receiver.c b/receiver.c
index e85c4779..1fcc2858 100644
--- a/receiver.c
+++ b/receiver.c
@@ -881,7 +881,7 @@ int recv_files(int f_in, int f_out, char *local_name)
                                        do_unlink(partialptr);
                                handle_partial_dir(partialptr, PDIR_DELETE);
                        }
-               } else if (keep_partial && partialptr && !one_inplace) {
+               } else if (keep_partial && partialptr && (!one_inplace || delay_updates)) {
                        if (!handle_partial_dir(partialptr, PDIR_CREATE)) {
                                rprintf(FERROR,
                                        "Unable to create partial-dir for %s -- discarding %s.\n",
```

**Comment by sthen** [comment]
Also seeing this with OpenBSD mirrors FWIW.
